### PR TITLE
[sagemaker][test]Multi region sm inherit

### PR DESCRIPTION
**Pull request by SergTogul**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Description

SM tests fail from time to time because of capacity. Added a functionality to rerun tests in different regions if they fail in one of them.
Refactored training tests for all frameworks. Refactored inference tests where needed. 

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true` or `neuron_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/sagemaker_tests/huggingface_pytorch/training/integration/sagemaker/test_smdp.py [hunk 1]
```diff
@@ -72,45 +75,18 @@ def get_transformers_version(ecr_image):
...
         raise LookupError("HF transformers version not found in image URI")
 
+
 @pytest.mark.integration("smdataparallel")
 @pytest.mark.model("hf_qa_smdp")
 @pytest.mark.processor("gpu")
 @pytest.mark.skip_cpu
 @pytest.mark.skip_py2_containers
-def test_smdp_question_answering(ecr_image, instance_type, py_version, sagemaker_session, tmpdir):
+def test_smdp_question_answering(ecr_image, sagemaker_regions, py_version, ):
```
Comment (tejaschumbalkar) [inline test/sagemaker_tests/huggingface_pytorch/training/integration/sagemaker/test_smdp.py:84]: nit: additional `, ` can be removed

File: test/sagemaker_tests/huggingface_pytorch/training/integration/sagemaker/test_smdp.py [hunk 2]
```diff
@@ -72,45 +75,18 @@ def get_transformers_version(ecr_image):
...
-            instance_count=instance_count,
-            instance_type=instance_type,
-            sagemaker_session=sagemaker_session,
-            py_version=py_version,
-            distribution=distribution,
-            hyperparameters=hyperparameters,
-        )
-        estimator.fit(job_name=sagemaker.utils.unique_name_from_base('test-hf-pt-qa-smdp'))
+    invoke_hf_pt_helper_function(ecr_image, sagemaker_regions, _test_smdp_question_answering_function,
+                                 py_version, 1)
```
Comment (tejaschumbalkar) [inline test/sagemaker_tests/huggingface_pytorch/training/integration/sagemaker/test_smdp.py:89]: function args should be used for better maintainability and readability
```c
function_args = {
            'py_version': py_version,
            'instances_quantity': 1,
        }
        
invoke_hf_pt_helper_function(ecr_image, sagemaker_regions, _test_smdp_question_answering_function, function_args)
```
and

File: test/sagemaker_tests/huggingface_pytorch/training/integration/sagemaker/test_smdp.py [hunk 3]
```diff
@@ -119,26 +95,29 @@ def test_smdp_question_answering(ecr_image, instance_type, py_version, sagemaker
 @pytest.mark.processor("gpu")
 @pytest.mark.skip_cpu
 @pytest.mark.skip_py2_containers
-def test_smdp_question_answering_multinode(ecr_image, instance_type, py_version, sagemaker_session, tmpdir):
+def test_smdp_question_answering_multinode(ecr_image, sagemaker_regions, py_version, ):
     """
     Tests SM Distributed DataParallel single-node via script mode
     """
+    invoke_hf_pt_helper_function(ecr_image, sagemaker_regions, _test_smdp_question_answering_function,
+                                 py_version, 2)
```
Comment (tejaschumbalkar) [inline test/sagemaker_tests/huggingface_pytorch/training/integration/sagemaker/test_smdp.py:103]: Same here, function args in the form of dict should be used instead of passing separate function arguments.

**Review by tejaschumbalkar — COMMENTED** [review]
function args in the form of dict should be used instead of passing separate function arguments. This needs to be addressed in all applicable test functions.

File: test/sagemaker_tests/huggingface_pytorch/training/integration/sagemaker/test_smdp.py (hunk 2 above)
Comment (SergTogul) [reply test/sagemaker_tests/huggingface_pytorch/training/integration/sagemaker/test_smdp.py:89]: As we agreed it will require quite a lot of work but won't add too much value since helper functions are literally 2 raws below.

File: test/sagemaker_tests/mxnet/training/__init__.py [hunk 4]
```diff
@@ -11,3 +11,39 @@
 # ANY KIND, either express or implied. See the License for the specific
 # language governing permissions and limitations under the License.
 from __future__ import absolute_import
+
+import sagemaker
+from sagemaker.mxnet.estimator import MXNet as MXNet_origin
+
+
+class MXNet(MXNet_origin):
```
Comment (tejaschumbalkar) [inline test/sagemaker_tests/mxnet/training/__init__.py:19]: Is there a reason we have implemented the MXNET retry in this way instead of the one added in TF? Also, Can we keep the mxnet estimator name as `MXNET` instead of `MXNet_origin`, we should rename the class name instead so that we can distinguish between a class and actual sagemaker estimator invocation

File: test/sagemaker_tests/mxnet/training/__init__.py [hunk 5]
```diff
@@ -11,3 +11,39 @@
...
+    def fit(self, inputs=None, wait=True, logs="All", job_name=None, experiment_config=None, **kwargs):
+        from ... import get_ecr_image_region, get_sagemaker_session, get_ecr_image
+
+        for region in self.sagemaker_regions:
+            self.sagemaker_session = get_sagemaker_session(region)
+            # Upload the image to test region if needed
+            if region != get_ecr_image_region(self.image_uri):
+                self.image_uri = get_ecr_image(self.image_uri, region)
+            try:
+                super().fit(
```
Comment (tejaschumbalkar) [inline test/sagemaker_tests/mxnet/training/__init__.py:36]: Does this actually work? The `MXNET` estimator is initiated with ECR image and SM session to the first region in `sagemaker_regions` list. On a ICE, the `super.fit(...)` api is called again but the `MXNET` estimator is not created/updated with the new ECR image and SM session? Does initiating `self.image_uri` and `self.sagemaker_session` do the job?

File: test/sagemaker_tests/autogluon/__init__.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    This function will rerun for all SM regions after a defined wait time if capacity issues are seen.
+
+    :param ecr_image: ECR image in us-west-2 region
+    :param sagemaker_regions: List of SageMaker regions
+    :param helper_function: Function to invoke
+    :param helper_function_args: Helper function args
+
+    :return: None
+    """
+    from .. import get_ecr_image_region, get_sagemaker_session, get_ecr_image
```
[the remark this answers is not in the document]
Comment (SergTogul) [reply test/sagemaker_tests/autogluon/__init__.py:30]: Sagemaker local tests fails if I declare it globally. Because only sagemaker remote copy .. folders.
I tried to fix local tests but it looked like way too much efforts

File: test/sagemaker_tests/pytorch/__init__.py [hunk 7]
```diff
@@ -175,21 +27,18 @@ def invoke_pytorch_helper_function(ecr_image, sagemaker_regions, helper_function
 
     :return: None
     """
+    from .. import get_ecr_image_region, get_sagemaker_session, get_ecr_image
```
[the remark this answers is not in the document]
Comment (SergTogul) [reply test/sagemaker_tests/pytorch/__init__.py:31]: I was trying to not change signatures. Don't add any more complexity into review

File: test/sagemaker_tests/mxnet/training/__init__.py (hunk 4 above)
Comment (SergTogul) [reply test/sagemaker_tests/mxnet/training/__init__.py:19]: MXNet tests are much simplier. Most of the time it's just _estimator.fit()_. I thought overwriting _fit()_ will safe a lot of efforts. 
Will rename the wrapper

File: test/sagemaker_tests/mxnet/training/__init__.py (hunk 5 above)
Comment (SergTogul) [reply test/sagemaker_tests/mxnet/training/__init__.py:36]: Correct. When you call `super.fit()` it uses current values for `self.image_uri` and `self.sagemaker_session`.

File: test/sagemaker_tests/huggingface_pytorch/training/integration/sagemaker/test_smmp.py [hunk 8]
```diff
@@ -67,30 +69,8 @@ def get_transformers_version(ecr_image):
...
-        volume_size=volume_size,
-        role="SageMakerRole",
-        image_uri=ecr_image,
-        distribution=distribution,
-        py_version=py_version,
-        hyperparameters=hyperparameters,
-        sagemaker_session=sagemaker_session,
-    )
-    huggingface_estimator.fit(job_name=sagemaker.utils.unique_name_from_base("test-hf-pt-qa-smmp"))
+def test_smmp_gpu(ecr_image, sagemaker_regions, instance_type, framework_version, py_version, dist_gpu_backend):
```
[the remark this answers is not in the document]
Comment (SergTogul) [reply test/sagemaker_tests/huggingface_pytorch/training/integration/sagemaker/test_smmp.py:72]: I was trying to not change signatures. Don't add any more complexity into review.

File: test/sagemaker_tests/pytorch/__init__.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (tejaschumbalkar) [reply test/sagemaker_tests/pytorch/__init__.py:31]: I pinned the question to unrelated location. Why is `num_retries` and `retry_delay` functionality removed from the retry logic?

File: test/sagemaker_tests/autogluon/__init__.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
```
Comment (tejaschumbalkar) [inline test/sagemaker_tests/autogluon/__init__.py:31]: Can we add `num_retries` and `retry_delay` to this functionality as implemented [here](https://github.com/aws/deep-learning-containers/blob/master/test/sagemaker_tests/pytorch/training/integration/sagemaker/__init__.py#L52)

Comment (tejaschumbalkar) [reply test/sagemaker_tests/autogluon/__init__.py:31]: Same with other `__init__.py` files

Comment (SergTogul) [reply test/sagemaker_tests/autogluon/__init__.py:31]: As we discussed, retried tests likely fail because of timeout. So I didn't add it.

File: test/sagemaker_tests/mxnet/training/integration/utils.py [hunk 10]
```diff
@@ -145,14 +124,15 @@ def reupload_image_to_test_ecr(source_image_uri, target_image_repo_name, target_
...
-    ECR_PASSWORD_FILE_PATH = os.path.join("/tmp", f"{get_unique_name_from_tag(source_image_uri)}.txt")
+    ecr_handler = EcrHandler()
     sts_client = boto3.client("sts", region_name=target_region)
     target_ecr_client = boto3.client("ecr", region_name=target_region)
     target_account_id = sts_client.get_caller_identity().get("Account")
     image_account_id = get_account_id_from_image_uri(source_image_uri)
     image_region = get_region_from_image_uri(source_image_uri)
     image_repo_uri, image_tag = source_image_uri.split(":")
     _, image_repo_name = image_repo_uri.split("/")
+    client = boto3.client('ecr', region_name = image_region)
```
Comment (arjkesh) [inline test/sagemaker_tests/mxnet/training/integration/utils.py:135]: thanks for making these changes

File: test/sagemaker_tests/tensorflow/tensorflow2_training/integration/sagemaker/test_smmodelparallel.py [hunk 11]
```diff
@@ -122,12 +139,20 @@ def test_smmodelparallel_multinode_efa(n_virginia_sagemaker_session, efa_instanc
 @pytest.mark.skip_cpu
 @pytest.mark.skip_py2_containers
 @pytest.mark.parametrize("test_script, num_processes", [("tf2_conv.py", 2), ("tf2_conv_xla.py", 2), ("smmodelparallel_hvd2_conv.py", 4), ("send_receive_checkpoint.py", 2), ("tf2_checkpoint_test.py", 2)])
-def test_smmodelparallel(n_virginia_sagemaker_session, instance_type, n_virginia_ecr_image, tmpdir, framework_version, test_script, num_processes):
+def test_smmodelparallel(ecr_image, sagemaker_regions, instance_type, tmpdir, framework_version, test_script, num_processes):
+    invoke_tf_helper_function(ecr_image,
```
Comment (arjkesh) [inline test/sagemaker_tests/tensorflow/tensorflow2_training/integration/sagemaker/test_smmodelparallel.py:143]: nit: let's run black on this, this seems a bit randomly formatted

File: test/sagemaker_tests/tensorflow/tensorflow2_training/integration/sagemaker/test_smmodelparallel.py [hunk 12]
```diff
@@ -156,12 +181,20 @@ def test_smmodelparallel(n_virginia_sagemaker_session, instance_type, n_virginia
 @pytest.mark.skip_cpu
 @pytest.mark.skip_py2_containers
 @pytest.mark.parametrize("test_script, num_processes", [("smmodelparallel_hvd2_conv_multinode.py", 2)])
-def test_smmodelparallel_multinode(n_virginia_sagemaker_session, instance_type, n_virginia_ecr_image, tmpdir, framework_version, test_script, num_processes):
+def test_smmodelparallel_multinode(ecr_image, sagemaker_regions, instance_type, tmpdir, framework_version, test_script, num_processes):
+    invoke_tf_helper_function(ecr_image,
```
Comment (arjkesh) [inline test/sagemaker_tests/tensorflow/tensorflow2_training/integration/sagemaker/test_smmodelparallel.py:185]: nit: let's run black on this, this seems a bit randomly formatted

File: test/sagemaker_tests/tensorflow/__init__.py [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+    This function will rerun for all SM regions after a defined wait time if capacity issues are seen.
+
+    :param ecr_image: ECR image in us-west-2 region
+    :param sagemaker_regions: List of SageMaker regions
+    :param helper_function: Function to invoke
+    :param helper_function_args: Helper function args
+
+    :return: None
+    """
+    from .. import get_ecr_image_region, get_sagemaker_session, get_ecr_image
```
Comment (arjkesh) [inline test/sagemaker_tests/tensorflow/__init__.py:33]: I'm a bit confused about this - why do we need to do a function level relative import here?

File: test/sagemaker_tests/tensorflow/__init__.py [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+
+def invoke_tf_helper_function(ecr_image, sagemaker_regions, helper_function, *helper_function_args):
+    """
+    Used to invoke SM job defined in the helper functions in respective test file. The ECR image and the sagemaker
+    session are passed explicitly depending on the AWS region.
+    This function will rerun for all SM regions after a defined wait time if capacity issues are seen.
+
+    :param ecr_image: ECR image in us-west-2 region
+    :param sagemaker_regions: List of SageMaker regions
+    :param helper_function: Function to invoke
```
Comment (arjkesh) [inline test/sagemaker_tests/tensorflow/__init__.py:28]: nit: This helper function seems like it needs to have. a very particular interface. Can you update the doctoring to include an example of a helper function, for instance?

File: test/sagemaker_tests/mxnet/inference/__init__.py [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+from __future__ import absolute_import
+
+import sagemaker
+
+
+
+def invoke_mxnet_helper_function(ecr_image, sagemaker_regions, helper_function, *helper_function_args):
```
Comment (arjkesh) [inline test/sagemaker_tests/mxnet/inference/__init__.py:19]: is there a reason we didn't define one for mxnet training? Also, is there any way we can make a shared utility instead of defining this on a framework level basis?

File: test/sagemaker_tests/tensorflow/__init__.py (hunk 13 above)
Comment (arjkesh) [reply test/sagemaker_tests/tensorflow/__init__.py:33]: I see the comment that there are some failures in SM local because of this. Noted

File: test/sagemaker_tests/autogluon/__init__.py [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+from __future__ import absolute_import
+import sagemaker
+
+def invoke_autogluon_helper_function(ecr_image, sagemaker_regions, helper_function, *helper_function_args):
```
[the remark this answers is not in the document]
Comment (arjkesh) [reply test/sagemaker_tests/autogluon/__init__.py:17]: Comment applies to all functions like this. I see that "sagemaker" is a dependency, though, so no worries if you want to keep it here.
